File: Computer Programming/C++/Recursion and Dynamic Programming/combinationSum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) 
    {
         int index = candidates.size()-1;
         vector<vector<int>> combo;
         vector<int>current;
         comb(combo,current,candidates,target,index);  

        return combo; 
    }

    void comb(vector<vector<int>>&combo,vector<int>current,vector<int>&candidates,int target,int index)
    {
        if(target==0)
        {
            combo.push_back(current);
            return;
        }
        if(index<0 || target<0)
        {
            return;
        }

        current.push_back(candidates[index]);
        comb(combo,current,candidates,target-candidates[index],index);
        current.pop_back();
        comb(combo,current,candidates,target,index-1);
        return;
    }
};

File: Computer Programming/Python/Recursion/Combination Sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """

        n = len(candidates)
        ans = []
        current = []
        self.function(ans,candidates,current,n-1,target)

        return ans

    def function(self,ans,candidates,current,index,target):
        if target < 0:
            return
        if target == 0:
            ans.append(list(current))
            return

        if index < 0:
            return

        current.append(candidates[index])
        self.function(ans,candidates,current,index,target - candidates[index])
        current.pop()

        self.function(ans,candidates,current,index-1,target)
        return
